module/zcommon/zfs_fletcher.c

```diff
@@ -245,11 +214,258 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
+
+	mutex_enter(&fletcher_4_impl_seletor_lock);
+	switch (idx) {
+	case FLETCHER_FASTEST:
+		fletcher_4_chosen = fletcher_4_fastest;
+		break;
+	case FLETCHER_ORIGINAL:
+	case FLETCHER_CYCLE:
+		fletcher_4_chosen = &fletcher_4_generic_calls;
+		break;
+	}
+
+	fletcher_4_impl_seletor = idx;
+	mutex_exit(&fletcher_4_impl_seletor_lock);
+
+	return (0);
+}
+
+static inline const struct fletcher_4_calls *fletcher_4_impl_get(void)
+{
+#if !defined(_KERNEL)
+	if (fletcher_4_impl_seletor == FLETCHER_CYCLE) {
+		static unsigned int cycle_count = 0;
+		const struct fletcher_4_calls *algo = NULL;
+		unsigned int index;
+
+		while (1) {
+			index = __sync_fetch_and_add(&cycle_count, 1);
+			algo = fletcher_4_algos[
+			    index % ZFS_ARRAY_SIZE(fletcher_4_algos)];
+			if (algo->valid == NULL || algo->valid())
+				break;
+		}
+		return (algo);
+	}
+#endif
+	asm volatile("":::"memory");
+	return (fletcher_4_chosen);
+}
+
+#define	BENCH_SIZE	4096
+
+void
+fletcher_4_init(void)
+{
+	static char databuf[BENCH_SIZE];
+	const uint64_t const bench_ns = (50 * 1000 * 1000ULL); /* 50ms */
+	kstat_named_t *stat;
+	unsigned long best_run_count = 0;
+	unsigned long best_run_index = 0;
+	int i;
+
+	stat = &fletcher_4_kstat_data[0];
+	for (i = 0; i < ZFS_ARRAY_SIZE(fletcher_4_algos); i++, stat++) {
+		const struct fletcher_4_calls *algo = fletcher_4_algos[i];
+		unsigned long run_count = 0;
+		hrtime_t start;
+		zio_cksum_t zc;
+
+		strncpy(stat->name, algo->name, sizeof (stat->name) - 1);
+		stat->data_type = KSTAT_DATA_UINT64;
+		stat->value.ui64 = 0;
+
+		if (algo->valid != NULL && !algo->valid())
+			continue;
+
+		kpreempt_disable();
+		start = gethrtime();
+		algo->init(&zc);
+		do {
+			algo->compute(databuf, sizeof (databuf), &zc);
+			run_count++;
+		} while (gethrtime() < start + bench_ns);
+		if (algo->fini != NULL)
+			algo->fini(&zc);
+		kpreempt_enable();
+
+		if (run_count > best_run_count) {
+			best_run_count = run_count;
+			best_run_index = i;
+		}
+
+		/*
+		 * Due to high overhead of gethrtime(), the performance data
+		 * here is inaccurate and much slower than it could be.
+		 * It's fine for our use though because only relative speed
+		 * is concerned.
+		 */
+		stat->value.ui64 = sizeof (databuf) * run_count *
+		    (NANOSEC / bench_ns) >> 20; /* by MB/s */
+	}
+
+	fletcher_4_fastest = fletcher_4_algos[best_run_index];
+
+	mutex_init(&fletcher_4_impl_seletor_lock, NULL, MUTEX_DEFAULT, NULL);
+	fletcher_4_impl_set("fastest");
+
+	fletcher_4_kstat = kstat_create("zcommon", 0, "fletcher_4_bench",
+	    "misc", KSTAT_TYPE_NAMED, ZFS_ARRAY_SIZE(fletcher_4_algos),
+	    KSTAT_FLAG_VIRTUAL);
+	if (fletcher_4_kstat != NULL) {
+		fletcher_4_kstat->ks_data = fletcher_4_kstat_data;
+		kstat_install(fletcher_4_kstat);
+	}
+}
+
+void
+fletcher_4_fini(void)
+{
+	mutex_destroy(&fletcher_4_impl_seletor_lock);
+	if (fletcher_4_kstat != NULL) {
+		kstat_delete(fletcher_4_kstat);
+		fletcher_4_kstat = NULL;
+	}
+}
+
 #if defined(_KERNEL) && defined(HAVE_SPL)
```

Comment:
As a matter of convention we want to the kernel module parameter blocks at the very end of the file.